File: 67. Add Binary.cpp
class Solution {
private:
    string res = "";
    int carry = 0;
public:
    string addBinary(string a, string b) {
        reverse(a.begin(), a.end());
        reverse(b.begin(), b.end());
        int maxlength = a.length() > b.length() ? a.length() : b.length();
        for (int i = 0; i < maxlength; i++)
        {
            if (i < a.length() && i < b.length())
                add(a[i], b[i]);
            else if (i < a.length() && i >= b.length())
                add(a[i], '0');
            else if (i >= a.length() && i < b.length())
                add('0', b[i]);
        }
        if (carry == 1)
            res = to_string(carry) + res;
        return res;
    }
    
    void add(char a, char b)
    {
        int temp = (a - '0') + (b - '0') + carry;
        if (temp >= 2)
        {
            res = to_string(temp - 2) + res;
            carry = 1;
        }
        else
        {
            res = to_string(temp) + res;
            carry = 0;
        }
    }
};

File: 67. Add Binary.py
class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        self.res = ""
        self.carry = 0
        a, b = a[::-1], b[::-1]
        for i in xrange(max(len(a), len(b))):
            if i < len(a) and i < len(b):
                self.add(a[i], b[i])
            elif i < len(a) and i>= len(b):
                self.add(a[i], 0)
            elif i < len(b) and i>= len(a):
                self.add(0, b[i])
        if self.carry == 1:
            self.res = str(self.carry) + self.res
        return self.res
        
    def add(self, a, b):
        temp = int(a) + int(b) + self.carry
        if temp >= 2:
            self.res = str(temp - 2) + self.res
            self.carry = 1
        else:
            self.res = str(temp) + self.res
            self.carry = 0
